### PR TITLE
More sophisticated migration tracking

### PR DESCRIPTION
**Pull request by SabreCat**
[//]: # (Note: See http://habitica.fandom.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # here, if applicable. This will automatically close the issue if your PR is merged in)

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)

1. Adds a `migrations` field to the user, deprecating `migration`.
2. Changes the routine `take-this.js` migration to use the new field, as an example:
* Trims existing data to a maximum of 5 entries in date order
* Uses the existence of a `migrations.name` value of the migration's name to check if the migration has already run on that user, instead of the single value of the `migration` field
* Saves information about the migration to the `migrations` field, in the form

```
migrations: [
  {
    name: '20190801_take_this',
    processed: new Date(),
    item: 'shield_special_takeThis', // arbitrary field name & data for troubleshooting
  },
  {
    name: '20190909_fix_things',
    processed: new Date(),
    fixed: 'party._id guilds',
  },
]
```

[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

### CODE REVIEW

File: migrations/users/take-this.js [hunk 1]
```diff
@@ -10,33 +11,50 @@ let count = 0;
 async function updateUser (user) {
   count++;
 
-  const set = {};
-  let push;
+  let set;
+
+  let migrationsArray = sortBy(user.migrations, ['processed']);
```
Comment (paglias) [inline migrations/users/take-this.js]: is the sortBy necessary with an array now? it's already sorted with the last one added at the end

File: website/server/models/user/schema.js [hunk 2]
```diff
@@ -73,7 +73,8 @@ let schema = new Schema({
   // We want to know *every* time an object updates. Mongoose uses __v to designate when an object contains arrays which
   // have been updated (http://goo.gl/gQLz41), but we want *every* update
   _v: { $type: Number, default: 0 },
-  migration: String,
+  migration: String, // deprecated
+  migrations: Array, // {name: '20190801_take_this', processed: new Date(), item: 'shield_special_takeThis'}
```
Comment (paglias) [inline website/server/models/user/schema.js]: can we specify this is an array of mixed objects? or if it always follow the `{name: '20190801_take_this', processed: new Date(), item: 'shield_special_takeThis'}` structure maybe define it?

**Review by paglias — COMMENTED** [review]
left a couple of comments but otherwise looks good

**Comment by paglias** [comment]
is this something we still want?

**Comment by SabreCat** [comment]
Ah yeah, might not be a bad idea! I can still clean this up if desired.

**Comment by SabreCat** [comment]
OK, how's that look?

**Review by paglias — COMMENTED** [review]
LGTM, I wonder if we could abstract a bit the pieces of code that are common to all migrations, for example in the migration runner script, to avoid having to remember to copy and paste code from an old migration.

**Comment by paglias** [comment]
@SabreCat I'm leaving this to you if you want to merge it

**Comment by SabreCat** [comment]
Going to close this for now--we haven't really had a need for this during the last several migrations we've run, so maybe it would complicate the schema for no noticeable benefit. We can always reopen this if it turns out we need it.

**Pull request closed without merging**